x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -354,6 +369,48 @@ public void authorizeIndexAction(RequestInfo requestInfo, AuthorizationInfo auth
         }
     }
 
+    private final Map<IndicesAccessControlCacheKey, ListenableFuture<IndicesAccessControl>> indicesAccessControlLookup =
+        new ConcurrentHashMap<>();
+
+    private void buildIndicesAccessControlAsync(
```

Comment:
This change will have an interplay with #78327 if both gets merged. In that case, my plan is to conditionally apply both optimisations: the optimisation here will only be picked if it is not covered by #78327, e.g. when the request does *not* target a single index.